src/main/java/com/gregtechceu/gtceu/api/recipe/ingredient/nbtpredicate/NBTPredicateUtils.java

```diff
@@ -0,0 +1,96 @@
+package com.gregtechceu.gtceu.api.recipe.ingredient.nbtpredicate;
+
+import net.minecraft.nbt.ByteTag;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.DoubleTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.StringTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.nbt.TagParser;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+
+public class NBTPredicateUtils {
+
+    public static JsonElement toJson(Tag tag) {
+        try {
+            return JsonParser.parseString(tag.getAsString());
+        } catch (Exception ignored) {
+            return new JsonPrimitive(tag.getAsString());
+        }
```

Comment:
this'll almost always return JsonPrimitives because SNBT has type specifiers that JSON doesn't.
Do this instead:
```suggestion
        return NbtOps.INSTANCE.convertTo(JsonOps.INSTANCE, tag);
```